From 069fa86274ba51a1504046f5e084e2c3fd6f2489 Mon Sep 17 00:00:00 2001
From: Lora Lin <lora.lin.wiwynn@gmail.com>
Date: Thu, 11 Jul 2024 10:10:27 +0800
Subject: [PATCH] Add event log type for PMIC error/VR alert

Summary:
- Add event log type for PMIC error/PMbus alert/VR fault event.

Change-Id: I4776b39ca345909bc2440c906d120741654374f9
Signed-off-by: Lora Lin <lora.lin.wiwynn@gmail.com>
Upstream-Status: Pending
---
 .../oem_meta_file_io_type_event_log.cpp       | 163 +++++++++++++++++-
 .../oem_meta_file_io_type_event_log.hpp       |   5 +
 2 files changed, 164 insertions(+), 4 deletions(-)

diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
index d063070..a02d95b 100644
--- a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
@@ -8,6 +8,8 @@ PHOSPHOR_LOG2_USING;
 namespace pldm::responder::oem_meta
 {
 
+uint8_t MAX_EVENT_LEN = 5;
+
 void EventLogHandler::parseObjectPathToGetSlot(uint64_t& slotNum)
 {
     static constexpr auto slotInterface =
@@ -96,13 +98,94 @@ int EventLogHandler::write(const message& eventData)
         " HDT_PRSNT Assertion",
         " PLTRST Assertion",
         " APML_ALERT Assertion",
+        " CXL1 HB",
+        " CXL2 HB",
         " Post-Started",
         " Post-Ended",
     });
 
+    static constexpr auto pmic_event_type = std::to_array({
+        " SWAout OV",
+        " SWBout OV",
+        " SWCout OV",
+        " SWDout OV",
+        " Vin Bulk OV",
+        " Vin Mgmt OV",
+        " SWAout UV",
+        " SWBout UV",
+        " SWCout UV",
+        " SWDout UV",
+        " Vin Bulk UV",
+        " Vin Mgmt To Vin Buck Switchover",
+        " High Temp Warning",
+        " Vout 1v8 PG",
+        " High Current Warning",
+        " Current Limit Warning",
+        " Critical Temp Shutdown",
+    });
+
+    static constexpr auto dimm_id = std::to_array({
+        " DIMM A0",
+        " DIMM A1",
+        " DIMM A2",
+        " DIMM A3",
+        " DIMM A4",
+        " DIMM A5",
+        " DIMM A6",
+        " DIMM A7",
+        " DIMM A8",
+        " DIMM A9",
+        " DIMM A10",
+        " DIMM A11",
+    });
+
+    static constexpr auto vr_source = std::to_array<VRSource>({
+        {" PVDDCR_CPU0", true},
+        {" PVDDCR_SOC", true},
+        {" PVDDCR_CPU1", true},
+        {" PVDDIO", true},
+        {" PVDD11_S3", true},
+        {" PVDDQ_AB_ASIC1", true},
+        {" P0V85_ASIC1", true},
+        {" PVDDQ_CD_ASIC1", true},
+        {" P0V8_ASIC1", true},
+        {" PVDDQ_AB_ASIC2", true},
+        {" P0V85_ASIC2", true},
+        {" PVDDQ_CD_ASIC2", true},
+        {" P0V8_ASIC2", true},
+        {" P1V5_RETIMER_1", false},
+        {" P0V9_STBY_1", false},
+        {" P3V3_E1S_0", false},
+        {" P3V3_E1S_1", false},
+        {" P12V_E1S_0", false},
+        {" P12V_E1S_1", false},
+        {" PVTT_AB_ASIC1", false},
+        {" PVTT_AB_ASIC2", false},
+        {" PVTT_CD_ASIC1", false},
+        {" PVTT_CD_ASIC2", false},
+        {" PVPP_AB_ASIC1", false},
+        {" PVPP_AB_ASIC2", false},
+        {" PVPP_CD_ASIC1", false},
+        {" PVPP_CD_ASIC2", false},
+    });
+
     /* 1. Get host slot number */
     uint64_t slot=0;
 
+    if (eventData.size() != MAX_EVENT_LEN)
+    {
+        error("{FUNC}: Invalid incoming data for event log, data size={SIZE}",
+              "FUNC", std::string(__func__), "SIZE", eventData.size());
+        return PLDM_ERROR;
+    }
+
+    if (eventData[0] >= eventList.size())
+    {
+        error("{FUNC}: Invalid event type for event log, event type={TYPE}",
+              "FUNC", std::string(__func__), "TYPE", eventData[0]);
+        return PLDM_ERROR;
+    }
+
     try
     {
         parseObjectPathToGetSlot(slot);
@@ -116,23 +199,95 @@ int EventLogHandler::write(const message& eventData)
 
     /* 2. Map event from list */
     std::string errorLog;
+    std::string event_name;
     uint8_t eventType = eventData[0];
     uint8_t eventStatus = eventData[1];
 
+    if (strcmp(eventList[eventType], " DIMM PMIC Error") == 0)
+    {
+        if ((eventData[2] >= dimm_id.size()) ||
+            (eventData[3] >= pmic_event_type.size()))
+        {
+            error("{FUNC}: Invalid dimm id={ID}, pmic event type={TYPE}",
+                  "FUNC", std::string(__func__), "ID", eventData[2], "TYPE",
+                  eventData[3]);
+            return PLDM_ERROR;
+        }
+
+        event_name = eventList[eventType] + std::string(" :") +
+                     dimm_id[eventData[2]] + pmic_event_type[eventData[3]];
+    }
+    else if ((strcmp(eventList[eventType], " PMALERT Assertion") == 0) ||
+             (strcmp(eventList[eventType], " VR Fault") == 0))
+    {
+        if (eventData[2] >= vr_source.size())
+        {
+            error("{FUNC}: Invalid vr source={VR_SOURCE}", "FUNC",
+                  std::string(__func__), "VR_SOURCE", eventData[2]);
+            return PLDM_ERROR;
+        }
+
+        if (vr_source[eventData[2]].isPMBus)
+        {
+            event_name = eventList[eventType] + std::string(" :") +
+                  vr_source[eventData[2]].netName + " status: 0x" +
+                  std::format("{:02x}", eventData[3]) +
+                  std::format("{:02x}", eventData[4]);
+        }
+        else
+        {
+            event_name = eventList[eventType] + std::string(" :") +
+                  vr_source[eventData[2]].netName;
+        }
+    }
+    else
+    {
+        event_name = eventList[eventType];
+    }
+
     if (eventStatus == EVENT_ASSERTED)
     {
-        errorLog = "Event: Host" + std::to_string(slot) + eventList[eventType] +
+        errorLog = "Event: Host" + std::to_string(slot) + event_name +
                    ", ASSERTED";
     }
     else if (eventStatus == EVENT_DEASSERTED)
     {
-        errorLog = "Event: Host" + std::to_string(slot) + eventList[eventType] +
+        errorLog = "Event: Host" + std::to_string(slot) + event_name +
                    ", DEASSERTED";
     }
 
     /* 3. Add SEL and journal */
-    lg2::error("{ERROR}", "ERROR", errorLog);   // Create log in journal
-    pldm::utils::reportError(errorLog.c_str()); // Create log on Dbus
+    constexpr auto loggingService = "xyz.openbmc_project.Logging";
+    constexpr auto loggingPath = "/xyz/openbmc_project/logging";
+    constexpr auto loggingInterface = "xyz.openbmc_project.Logging.Create";
+    if (event_name == " CXL1 HB" || event_name == " CXL2 HB")
+    {
+        lg2::info("{INFO}", "INFO", errorLog); // Create log in journal
+        auto& bus = pldm::utils::DBusHandler::getBus();
+        using EntrySeverity = sdbusplus::xyz::openbmc_project::Logging::server::Entry::Level;
+
+        try
+        {
+            auto severityStr =
+                sdbusplus::xyz::openbmc_project::Logging::server::convertForMessage(
+                    EntrySeverity::Informational);
+
+            auto method = bus.new_method_call(loggingService, loggingPath, loggingInterface, "Create");
+
+            std::map<std::string, std::string> addlData{};
+            method.append(errorLog.c_str(), severityStr, addlData);
+            bus.call_noreply(method, dbusTimeout);
+        }
+        catch (const std::exception& e)
+        {
+            error("Failed to create info-level D-Bus log: {ERR}", "ERR", e);
+        }
+    }
+    else
+    {
+        lg2::error("{ERROR}", "ERROR", errorLog); // Create log in journal
+        pldm::utils::reportError(errorLog.c_str()); // Create log on Dbus
+    }
 
     return PLDM_SUCCESS;
 }
diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.hpp b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.hpp
index e58c1a2..43843f4 100644
--- a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.hpp
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.hpp
@@ -12,6 +12,11 @@ enum eventAssert : uint8_t
     EVENT_ASSERTED = 0x01,
 };
 
+struct VRSource {
+    std::string netName;
+    bool isPMBus;
+};
+
 /** @class EventLogHandler
  *
  *  @brief Inherits and implements FileHandler. This class is used
-- 
2.25.1

